### PR TITLE
knapsack added

### PR DESCRIPTION
**Pull request by Parth20GitHub**
#716 

## Description

knapsack added 

## Type of PR

- [X] Feature enhancement

## Checklist
- [X] I have performed a self-review of my code.
- [X] I have read and followed the Contribution Guidelines.
- [X] I have tested the changes thoroughly before submitting this pull request.
- [X] I have provided relevant issue numbers, screenshots, and videos after making the changes.
- [X] I have commented my code, particularly in hard-to-understand areas.

**Resources for DSA Features on Our Site:**

- Tabs: [Switch Languages](https://www.codeharborhub.live/community/features#multi-language-support-code-blocks)
- [Line Numbering](https://www.codeharborhub.live/community/features#line-numbering)
- [Math Equations](https://www.codeharborhub.live/community/features#math-equations-with-katex)

### CODE REVIEW

**Comment by Parth20GitHub** [comment]
please add GSSOC label with level

File: dsa/Algorithms/knapsack.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+Input: N = 3, W = 4, profit[] = {1, 2, 3}, weight[] = {4, 5, 1}
+Output: 3
+
+Explanation: There are two items which have weight less than or equal to 4. If we select the item with weight 4, the possible profit is 1. And if we select the item with weight 1, the possible profit is 3. So the maximum possible profit is 3. Note that we cannot put both the items with weight 4 and 1 together as the capacity of the bag is 4.
+
+Input: N = 3, W = 3, profit[] = {1, 2, 3}, weight[] = {4, 5, 6}
+Output: 0
+
+## Java Implementation
+
```
Comment (Ajay-Dhangar) [inline dsa/Algorithms/knapsack.md:31]: ### Bad Practice

```mdx

class Knapsack {

    // A utility function that returns
    // maximum of two integers
    static int max(int a, int b) { return (a > b) ? a : b; }

    // Returns the maximum value that
    // can be put in a knapsack of
    // capacity W
    static int knapSack(int W, int wt[], int val[], int n)
    {
        // Base Case
        if (n == 0 || W == 0)
            return 0;

        // If weight of the nth item is
        // more than Knapsack capacity W,
        // then this item cannot be included
        // in the optimal solution
        if (wt[n - 1] > W)
            return knapSack(W, wt, val, n - 1);

        // Return the maximum of two cases:
        // (1) nth item included
        // (2) not included
        else
            return max(val[n - 1]
                           + knapSack(W - wt[n - 1], wt,
                                      val, n - 1),
                       knapSack(W, wt, val, n - 1));
    }

    // Driver code
    public static void main(String args[])
    {
        int profit[] = new int[] { 60, 100, 120 };
        int weight[] = new int[] { 10, 20, 30 };
        int W = 50;
        int n = profit.length;
        System.out.println(knapSack(W, weight, profit, n));
    }
}
```

### Good Practice

````mdx
```java

class Knapsack {

    // A utility function that returns
    // maximum of two integers
    static int max(int a, int b) { return (a > b) ? a : b; }

    // Returns the maximum value that
    // can be put in a knapsack of
    // capacity W
    static int knapSack(int W, int wt[], int val[], int n)
    {
        // Base Case
        if (n == 0 || W == 0)
            return 0;

        // If weight of the nth item is
        // more than Knapsack capacity W,
        // then this item cannot be included
        // in the optimal solution
        if (wt[n - 1] > W)
            return knapSack(W, wt, val, n - 1);

        // Return the maximum of two cases:
        // (1) nth item included
        // (2) not included
        else
            return max(val[n - 1]
                           + knapSack(W - wt[n - 1], wt,
                                      val, n - 1),
                       knapSack(W, wt, val, n - 1));
    }

    // Driver code
    public static void main(String args[])
    {
        int profit[] = new int[] { 60, 100, 120 };
        int weight[] = new int[] { 10, 20, 30 };
        int W = 50;
        int n = profit.length;
        System.out.println(knapSack(W, weight, profit, n));
    }
}
```
````

File: dsa/Algorithms/knapsack.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        int profit[] = new int[] { 60, 100, 120 };
+        int weight[] = new int[] { 10, 20, 30 };
+        int W = 50;
+        int n = profit.length;
+        System.out.println(knapSack(W, weight, profit, n));
+    }
+}
+
+### Time Complexity: O(2N)
```
Comment (Ajay-Dhangar) [inline dsa/Algorithms/knapsack.md:75]: replace `O(2N)` to `$O(2N)$` and `O(2N)` to `$O(2N)$`

**Review by Ajay-Dhangar — CHANGES_REQUESTED** [review]
now update your code based on suggestions and you need to read our contribution guidelines and resources for How to write code ? because you can't follow our structure so, next time you need to follow it.

File: dsa/Algorithms/knapsack.md (hunk 2 above)
Comment (Parth20GitHub) [reply dsa/Algorithms/knapsack.md:75]: ok I will do it

**Comment by Parth20GitHub** [comment]
I have updated it, please label it as level-2

**Review by Ajay-Dhangar — APPROVED** [review]
Done

**Comment by Parth20GitHub** [comment]
please update the leaderboard

**Pull request merged**